File: sample-core/src/main/java/com/quickblox/sample/core/async/BaseAsyncTask.java
package com.quickblox.sample.core.async;

import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

public abstract class BaseAsyncTask<Params, Progress, Result> extends AsyncTask<Params, Progress, Result> {
    private static final String TAG = BaseAsyncTask.class.getSimpleName();

    private static final Handler mainThreadHandler = new Handler(Looper.getMainLooper());
    private boolean isExceptionOccurred;

    @Override
    protected final Result doInBackground(Params... params) {
        try {
            return performInBackground(params);
        } catch (final Exception e) {
            isExceptionOccurred = true;
            mainThreadHandler.post(new Runnable() {
                @Override
                public void run() {
                    onException(e);
                }
            });
            return null;
        }
    }

    @Override
    protected final void onPostExecute(Result result) {
        if (!isExceptionOccurred) {
            onResult(result);
        }
    }

    public abstract Result performInBackground(Params... params) throws Exception;

    public abstract void onResult(Result result);

    public void onException(Exception e) {
        Log.w(TAG, e);
    }
}

File: sample-chat-kotlin/app/src/main/java/com/quickblox/sample/chat/kotlin/async/BaseAsyncTask.kt
package com.quickblox.sample.chat.kotlin.async

import android.os.AsyncTask
import android.os.Handler
import android.os.Looper
import android.util.Log


abstract class BaseAsyncTask<Params, Progress, Result> : AsyncTask<Params, Progress, Result>() {
    private val TAG = BaseAsyncTask::class.java.simpleName

    private val mainThreadHandler = Handler(Looper.getMainLooper())
    private var isExceptionOccurred: Boolean = false

    override fun doInBackground(vararg params: Params): Result? {
        try {
            return performInBackground(*params)
        } catch (e: Exception) {
            isExceptionOccurred = true
            mainThreadHandler.post { onException(e) }
            return null
        }
    }

    override fun onPostExecute(result: Result) {
        if (!isExceptionOccurred) {
            onResult(result)
        }
    }

    @Throws(Exception::class)
    abstract fun performInBackground(vararg params: Params): Result?

    abstract fun onResult(result: Result?)

    open fun onException(e: Exception) {
        Log.w(TAG, e)
    }
}